File: src/backend/internal/model/report_provenance.go
package model

import (
	"encoding/json"
	"time"
)

type ReportChartBinding struct {
	ChartKey     string            `json:"chart_key"`
	Title        string            `json:"title"`
	QueryKeys    []string          `json:"query_keys"`
	Formula      string            `json:"formula"`
	FieldMapping map[string]string `json:"field_mapping"`
}

// 聚合执行记录为不可变快照，不存储用户明细。
type ReportQueryExecution struct {
	ID             string               `json:"id"`
	ReportID       string               `json:"report_id"`
	SourceID       string               `json:"source_id"`
	APIName        string               `json:"api_name"`
	SourceName     string               `json:"source_name"`
	QueryKey       string               `json:"query_key"`
	RequestJSON    json.RawMessage      `json:"request_json"`
	SQL            string               `json:"sql"`
	QueryID        string               `json:"query_id"`
	Status         string               `json:"status"`
	ErrorMessage   string               `json:"error_message"`
	Rows           []map[string]any     `json:"rows"`
	CreatedAt      time.Time            `json:"created_at"`
	DurationMS     int64                `json:"duration_ms"`
	StartDate      string               `json:"start_date"`
	EndDate        string               `json:"end_date"`
	Cities         []string             `json:"cities"`
	BindingVersion string               `json:"binding_version"`
	Bindings       []ReportChartBinding `json:"bindings"`
	ReplayOf       string               `json:"replay_of,omitempty"`
}

type ReportProvenance struct {
	Bindings   []ReportChartBinding   `json:"bindings"`
	Executions []ReportQueryExecution `json:"executions"`
	Available  bool                   `json:"available"`
	Message    string                 `json:"message"`
}

type ReportProvenanceReplay struct {
	Execution  *ReportQueryExecution `json:"execution"`
	ReplayOnly bool                  `json:"replay_only"`
	Message    string                `json:"message"`
}

File: src/frontend/src/api/reportProvenance.ts
import { get, post } from './client';

export interface ChartBinding {
  chart_key: string;
  title: string;
  query_keys: string[];
  formula: string;
  field_mapping: Record<string, unknown>;
}
export interface ChartExecution {
  id: string;
  report_id: string;
  source_id: string;
  source_name?: string;
  api_name?: string;
  query_key: string;
  request_json: Record<string, unknown>;
  sql: string;
  query_id: string;
  status: string;
  error_message: string;
  rows: Record<string, unknown>[];
  created_at: string;
  start_date: string;
  end_date: string;
  cities: string[];
  bindings: ChartBinding[];
}
export interface ChartProvenance {
  available: boolean;
  message: string;
  bindings: ChartBinding[];
  executions: ChartExecution[];
}
export const executionsForChart = (executions: ChartExecution[], chartKey: string) => executions.filter((execution) =>
  execution.bindings?.some((binding) => binding.chart_key === chartKey && binding.query_keys.includes(execution.query_key)));
export function currentChartExecutions(executions: ChartExecution[], ids: string[], chartKey: string, start?: string, end?: string, dataDate?: string) {
  const snapshotOnly = chartKey === 'all_distribution' || chartKey === 'order_distribution';
  const from = snapshotOnly ? dataDate : start;
  const to = snapshotOnly ? dataDate : end;
  return executions.filter((item) => ids.includes(item.id) && (!snapshotOnly || Boolean(dataDate))
    && (!from || item.end_date >= from) && (!to || item.start_date <= to));
}
export const getChartProvenance = (id: string, executionIds: string[] = []) => {
  const params = new URLSearchParams();
  [...new Set(executionIds)].slice(0, 500).forEach((executionId) => params.append('execution_id', executionId));
  return get<ChartProvenance>(`/api/reports/${encodeURIComponent(id)}/provenance?${params.toString()}`);
};
export const replayChartExecution = (id: string, executionId: string) => post<{
  execution: ChartExecution; replay_only: boolean; message: string;
}>(`/api/reports/${encodeURIComponent(id)}/provenance/replay`, { execution_id: executionId });
